File: src/impl/Combination.cpp
#include <vector>
#include <format>
#include <iostream>
#include <polyfill/generator.hpp>
#include <polyfill/ranges.hpp>
#include <random>
#include <ranges>
#include <unordered_map>
#include <unordered_set>

#include <impl/Combination.hpp>

namespace mynum::impl {

static mynum::polyfill::generator<int> submasks(int mask) {
  int submask = mask;
  while (submask != 0) {
    submask = (submask - 1) & mask;
    co_yield submask;
  }
}

static mynum::polyfill::generator<std::pair<int, int>> submask_complement_pairs(int mask, bool symmetry = false) {
  for (auto submask : submasks(mask)) {
    auto complement = mask ^ submask;
    if (!symmetry || submask < complement) {
      co_yield {submask, complement};
    }
  }
}

std::ostream &operator<<(std::ostream &os, const Combination &comb) {
  return os << std::format("Combination({}, {{ {}, {}, {}, {}, {}, {} }}", comb.target, comb.numbers[0],
                           comb.numbers[1], comb.numbers[2], comb.numbers[3], comb.numbers[4], comb.numbers[5]);
}

Combination Combination::generate() {
  static std::random_device rd;
  static std::seed_seq ssq{rd()};
  static std::mt19937 gen{ssq};

  static std::uniform_int_distribution target_dist(1, 999);
  static std::uniform_int_distribution digits_dist(1, 9);

  static std::uniform_int_distribution middle_dist(0, 2);
  static std::array middle_numbers = {10, 15, 20};

  static std::uniform_int_distribution large_dist(0, 3);
  static std::array large_numbers = {25, 50, 75, 100};

  auto target = target_dist(gen);
  std::array<int, 6> numbers = {
      digits_dist(gen),
      digits_dist(gen),
      digits_dist(gen),
      digits_dist(gen),
      middle_numbers[middle_dist(gen)],
      large_numbers[large_dist(gen)],
  };

  return {target, numbers};
}

std::vector<std::shared_ptr<StateValue>> Combination::allSolutions() const {
  std::vector<std::unordered_map<int, std::vector<std::shared_ptr<StateValue>>>> states(1ull << numbers.size());

  for (const auto &[i, v] : mynum::compat::views::enumerate(numbers)) {
    states[1ull << i][v].push_back(std::make_shared<StateValue>(v, Operator::NONE, nullptr, nullptr));
  }

  std::vector<std::shared_ptr<StateValue>> solutions;
  std::unordered_set<std::string> seenHashes;

  for (int mask : std::views::iota(1, 1 << numbers.size())) {
    for (auto [submask1, submask2] : submask_complement_pairs(mask, false)) {
      auto state_vec_pairs = mynum::compat::views::cartesian_product(mynum::compat::views::values(states[submask1]),
                                                                    mynum::compat::views::values(states[submask2]));

      for (const auto &[state_vec1, state_vec2] : state_vec_pairs) {
        auto state_pairs =
            mynum::compat::views::cartesian_product(state_vec1, state_vec2) |
            mynum::compat::views::filter([](const auto &pair) { return std::get<0>(pair)->value < std::get<1>(pair)->value; });

        for (const auto &[state1, state2] : state_pairs) {
          std::vector<std::shared_ptr<StateValue>> new_states;
          new_states.reserve(6);
          new_states.push_back(StateValue::combine(state1, state2, Operator::ADD));
          new_states.push_back(StateValue::combine(state1, state2, Operator::MUL));
          new_states.push_back(StateValue::combine(state2, state1, Operator::SUB));
          if (state2->value != 0 && state1->value % state2->value == 0) {
            new_states.push_back(StateValue::combine(state1, state2, Operator::DIV));
          }
          if (state1->value != 0 && state2->value % state1->value == 0) {
            new_states.push_back(StateValue::combine(state2, state1, Operator::DIV));
          }

          for (auto new_state : new_states) {
            auto str = new_state->reconstruct();
            if (seenHashes.contains(str)) {
              continue;
            }

            states[mask][new_state->value].push_back(new_state);
            seenHashes.insert(str);
            if (target - new_state->value == 0) {
              solutions.push_back(new_state);
            }
          }
        }
      }
    }
  }

  return solutions;
}

std::shared_ptr<StateValue> Combination::solve() const {
  std::vector<std::unordered_map<int, std::shared_ptr<StateValue>>> states(1ull << numbers.size());

  for (const auto &[i, v] : mynum::compat::views::enumerate(numbers)) {
    states[1ull << i][v] = std::make_shared<StateValue>(v, Operator::NONE, nullptr, nullptr);
  }

  auto best_diff = std::numeric_limits<unsigned int>::max();
  std::shared_ptr<StateValue> best_match;

  for (int mask : std::views::iota(1, 1 << numbers.size())) {
    for (auto [submask1, submask2] : submask_complement_pairs(mask)) {
      auto state_pairs = mynum::compat::views::cartesian_product(mynum::compat::views::values(states[submask1]),
                                                                 mynum::compat::views::values(states[submask2]));

      for (const auto &[state1, state2] : state_pairs) {
        auto val1 = state1->value;
        auto val2 = state2->value;

        std::vector<std::shared_ptr<StateValue>> new_states;
        new_states.reserve(6);
        new_states.push_back(StateValue::combine(state1, state2, Operator::ADD));
        new_states.push_back(StateValue::combine(state1, state2, Operator::MUL));
        if (val1 > val2) {
          new_states.push_back(StateValue::combine(state1, state2, Operator::SUB));
        }
        if (val2 > val1) {
          new_states.push_back(StateValue::combine(state2, state1, Operator::SUB));
        }
        if (val2 != 0 && val1 % val2 == 0) {
          new_states.push_back(StateValue::combine(state1, state2, Operator::DIV));
        }
        if (val1 != 0 && val2 % val1 == 0) {
          new_states.push_back(StateValue::combine(state2, state1, Operator::DIV));
        }

        for (auto new_state : new_states) {
          if (states[mask].contains(new_state->value)) {
            new_state.reset();
            continue;
          }

          states[mask][new_state->value] = new_state;

          unsigned int diff = std::abs(target - new_state->value);

          if (diff >= best_diff) {
            continue;
          }

          best_diff = diff;
          best_match = new_state;

          if (diff == 0) {
            return new_state;
          }
        }
      }
    }
  }

  return best_match;
}

} // namespace mynum::impl

File: src/js-fallback/Combination.ts
import type { ICombination } from './types';
import { Solution } from './Solution';
import { OperatorEnum, StateValue } from './StateValue';
import { cartesian2, randomInt, range, submask_complement_pairs } from './Util';

const MIDDLE_NUMBERS = [10, 15, 20] as const;
const LARGE_NUMBERS = [25, 50, 75, 100] as const;

function copyNumbers(numbers: ICombination['numbers']): ICombination['numbers'] {
  return [
    numbers[0] ?? 0,
    numbers[1] ?? 0,
    numbers[2] ?? 0,
    numbers[3] ?? 0,
    numbers[4] ?? 0,
    numbers[5] ?? 0,
  ];
}

export class Combination implements ICombination {
  target: number = 0;
  numbers: ICombination['numbers'] = [0, 0, 0, 0, 0, 0];

  constructor(targetOrConfig?: number | ICombination, numbers?: ICombination['numbers']) {
    const argc = arguments.length;

    if (argc === 0) {
      return;
    }

    if (argc === 1 && typeof targetOrConfig === 'object' && targetOrConfig !== null) {
      if (typeof targetOrConfig.target !== 'number' || !Array.isArray(targetOrConfig.numbers)) {
        throw new TypeError('Invalid arguments.');
      }
      this.target = targetOrConfig.target;
      this.numbers = copyNumbers(targetOrConfig.numbers);
      return;
    }

    if (argc === 2) {
      if (typeof targetOrConfig !== 'number' || !Array.isArray(numbers)) {
        throw new TypeError('Invalid arguments.');
      }
      this.target = targetOrConfig;
      this.numbers = copyNumbers(numbers);
      return;
    }

    throw new TypeError('Too many arguments');
  }

  static generate(): Combination {
    const target = randomInt(1, 999);
    const numbers: ICombination['numbers'] = [
      randomInt(1, 9),
      randomInt(1, 9),
      randomInt(1, 9),
      randomInt(1, 9),
      MIDDLE_NUMBERS[randomInt(0, MIDDLE_NUMBERS.length - 1)],
      LARGE_NUMBERS[randomInt(0, LARGE_NUMBERS.length - 1)],
    ];
    return new Combination(target, numbers);
  }

  solve(): Solution | null {
    const states: Map<number, StateValue>[] = Array.from({ length: 64 }, () => new Map());

    for (const i of range(0, 6)) {
      states[1 << i].set(
        this.numbers[i],
        new StateValue(this.numbers[i], OperatorEnum.NONE, null, null),
      );
    }

    let bestDiff = Number.POSITIVE_INFINITY;
    let bestMatch: StateValue | null = null;

    for (const mask of range(1, 64)) {
      for (const [submask1, submask2] of submask_complement_pairs(mask)) {
        const statePairs = cartesian2(states[submask1].values(), states[submask2].values());

        for (const [state1, state2] of statePairs) {
          const val1 = state1.value;
          const val2 = state2.value;
          const newStates: StateValue[] = [
            StateValue.combine(state1, state2, OperatorEnum.ADD),
            StateValue.combine(state1, state2, OperatorEnum.MUL),
          ];

          if (val1 > val2) {
            newStates.push(StateValue.combine(state1, state2, OperatorEnum.SUB));
          }
          if (val2 > val1) {
            newStates.push(StateValue.combine(state2, state1, OperatorEnum.SUB));
          }
          if (val2 !== 0 && val1 % val2 === 0) {
            newStates.push(StateValue.combine(state1, state2, OperatorEnum.DIV));
          }
          if (val1 !== 0 && val2 % val1 === 0) {
            newStates.push(StateValue.combine(state2, state1, OperatorEnum.DIV));
          }

          for (const newState of newStates) {
            if (states[mask].has(newState.value)) {
              continue;
            }

            states[mask].set(newState.value, newState);

            const diff = Math.abs(this.target - newState.value);
            if (diff >= bestDiff) {
              continue;
            }

            bestDiff = diff;
            bestMatch = newState;
            if (diff === 0) {
              return new Solution(newState);
            }
          }
        }
      }
    }

    return bestMatch ? new Solution(bestMatch) : null;
  }

  toString(): string {
    return `Combination { target: ${this.target}, numbers : [${this.numbers.join(', ')}] }`;
  }

  toJSON(): ICombination {
    return {
      target: this.target,
      numbers: this.numbers,
    };
  }
}
